core/schema/host.go

```diff
@@ -243,6 +272,499 @@ type hostDirectoryArgs struct {
+				{
+					Name:  "path",
+					Value: dagql.NewString(path),
+				},
+			},
+		},
+		{
+			Field: "asGit",
+		},
+	}
+}
+
+func queryLocalGitRef(path, ref string) []dagql.Selector {
+	q := queryLocalGitRepository(path)
+	if ref == "" {
+		q = append(q, dagql.Selector{
+			Field: "head",
+		})
+	} else {
+		q = append(q, dagql.Selector{
+			Field: "ref",
+			Args: []dagql.NamedInput{
+				{
+					Name:  "name",
+					Value: dagql.NewString(ref),
+				},
+			},
+		})
+	}
+	return q
+}
+
+func queryRemoteGitRepository(gitURL *gitutil.GitURL) []dagql.Selector {
+	return []dagql.Selector{
+		{
+			Field: "git",
+			Args: []dagql.NamedInput{
+				{
+					Name:  "url",
+					Value: dagql.NewString(gitURL.Remote()),
+				},
+			},
+		},
+	}
+}
+
+func (s *hostSchema) resourceAsGitRef(
+	ctx context.Context,
+	r dagql.ObjectResult[*core.HostResource],
+	args struct{},
+) (
+	inst dagql.ObjectResult[*core.GitRef],
+	err error,
+) {
+	var (
+		addr = r.Self().Address
+		q    []dagql.Selector
+	)
+	gitURL, err := gitutil.ParseURL(addr)
+	if err == nil {
+		// Remote ref
+		if gitURL.Fragment != nil && gitURL.Fragment.Subdir != "" {
+			return inst, fmt.Errorf("git ref address cannot contain subdir")
+		}
+		q = queryRemoteGitRef(gitURL)
+	} else {
+		// Local ref
+		path, ref, _ := strings.Cut(addr, "#")
+		path, err = getLocalPath(path)
+		if err != nil {
+			return inst, err
+		}
+		q = queryLocalGitRef(path, ref)
+	}
+	err = s.srv.Select(ctx, s.srv.Root(), &inst, q...)
+	return inst, err
+}
+
+func (s *hostSchema) resourceAsSecret(
+	ctx context.Context,
+	r dagql.ObjectResult[*core.HostResource],
+	args struct{},
+) (
+	inst dagql.ObjectResult[*core.Secret],
+	err error,
+) {
+	var (
+		uri      string
+		cacheKey string
+	)
+	addr := r.Self().Address
+	if !strings.Contains(addr, ":") {
+		// case of e.g. `MY_ENV_SECRET`, which is shorthand for `env://MY_ENV_SECRET`
+		addr = "env://" + addr
+	}
+	// legacy secrets in the form of `--token env:MY_ENV_SECRET` instead of `env://MY_ENV_SECRET`
+	secretSource, val, _ := strings.Cut(addr, ":")
+	if !strings.HasPrefix(val, "//") {
+		addr = secretSource + "://" + val
+	}
+
+	addrWithoutQuery, queryValsStr, ok := strings.Cut(addr, "?")
+	if ok && len(queryValsStr) > 0 {
+		queryVals, err := url.ParseQuery(queryValsStr)
+		if err != nil {
+			return inst, err
+		}
+		if cacheKey := queryVals.Get("cacheKey"); cacheKey != "" {
+			cacheKey = cacheKey
+			queryVals.Del("cacheKey")
+			queryValsStr = queryVals.Encode()
+			if len(queryValsStr) > 0 {
+				addr = fmt.Sprintf("%s?%s", addrWithoutQuery, queryValsStr)
+			} else {
+				addr = addrWithoutQuery
+			}
+		}
+	}
+	uri = addr
+	err = s.srv.Select(ctx, s.srv.Root(), &inst, dagql.Selector{
+		Field: "secret",
+		Args: []dagql.NamedInput{
+			{
+				Name:  "uri",
+				Value: dagql.NewString(uri),
+			},
+			{
+				Name:  "cacheKey",
+				Value: dagql.Opt(dagql.String(cacheKey)),
+			},
+		},
+	})
+	return inst, err
+}
+
+func (s *hostSchema) resourceAsService(
```

Comment:
Thoughts on a future `resourceAsContainer`? That would be used for parsing the CLI args for a remote image.

Reason I'm raising, is because in the context of https://github.com/dagger/dagger/issues/10737, there's two potential places to load from: the host, or a remote registry.

Reply:
This PR already defines `resourceAsContainer`, is that what you mean? https://github.com/dagger/dagger/pull/10753/files/2cc8e15e087cb659a3f919892587a15206927976#diff-09f798f10567cb9017b829bfa481df0a90e444368d36bd8f7025a69d1d2521c9R486

Reply:
Ah, it's at the top :woman_facepalming: Yeah that!

Just kinda wondering how this would work if a host container store was connected, it should probably be able to look there as well?